packages/backend-common/src/errors.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * A set of common business logic errors.
+ *
+ * The error handler middleware understands these and will translate them to
+ * well formed HTTP responses.
+ *
+ * While these are intentionally analogous to HTTP errors, they are not
+ * intended to be thrown by the request handling layer. In those places, please
+ * use e.g. the http-errors library.
+ */
+
+class CustomErrorBase extends Error {
+  readonly cause?: Error;
+
+  constructor(constructor: Function, message?: string, cause?: Error) {
+    super(message);
+    Object.setPrototypeOf(this, constructor.prototype);
+    Error.captureStackTrace(this, constructor);
+    this.name = this.constructor.name;
+    this.cause = cause;
+  }
+
+  toString() {
```

Comment:
I think passing the message created here on to super would be better. With this you'll get different behaviour depending on how the error is used. For example the stack trace won't include the cause, or someone using `error.message` to print the error without the constructor name prefix.

Reply:
Is it ok to "mess with" the error's message like that? Sure I could change that.

Oooor should I remove the cause functionality?

Reply:
I mean you're already messing with it? :p

Yeah we could skip it for now, but I usually end up doing it manually in the message, and losing information in the process since only the message of the cause is kept.

Reply:
Chose to retain it, yeah it can be useful